### ISSUE TITLE
All APIs that use post are broken

### ISSUE THREAD

**Issue by pdath** [tejashah88/node-meraki-dashboard#10]
All APIs that use rest.post (aka a POST request to the Meraki API as opposed to a GET request) fail with a "TypeError: Cannot convert undefined or null to object" error.

A sample API to reproduce the problem is:
dashboard.devices.reboot("L_610800699462******","****-****-****");

I have tracked the issue down to a bug in axios.
https://github.com/axios/axios/issues/2460
Basically if you do a POST and the server redirects axios is incorrectly changing it to a GET request.

To further confirm the issue if you modify the base URL from https://api.meraki.com/api/v0/ to the URL of the shard that the organisation is hosted on the issue does not happen.

Apart from the axios bug what this reveals is each API call is probably taking almost twice as long as they could due to network latency.  First axios makes a request to api.meraki.com, gets a redirect, and then makes the call again to the shard URL.

Any potential fix needs to be carefull to make sure it still works when using multiple organisations at the same time.

**Comment by tejashah88** [comment]
@pdath Thank you for bringing this to my attention! I just saw your post on the Meraki forums. I agree that any fix that's done will have to be carefully laid out, as this is essentially a breaking bug that seems to be designed as such.

There's two things that can be done:
- Add a method that allows you to change the base url
- Enforces that the correct request operation is being made. This would be in the form of either making a fork of axios and relevant libraries to fix this behaviour or manually applying a "monkey patch" for the relevant libraries until. A PR could be made to address this bug for the respective libraries.

**Comment by tejashah88** [comment]
I'll take some time next week to write a few unit tests to document this bug and figure out how exactly this bug can be fixed for now.

**Comment by pdath** [comment]
I had a go at adding a method to change the base url.  The problem is you dont know what baseurl to use until after you have made at least one API call.  axios.create() gets called before any function executes, so you can't seem to change it afterwards (or at least I could not).

I also had a play with changing both dataProcessor and get() and trying to update the baseURL with response.request.socket.servername.  So basically the last request is used to update the baseURL for the next request.  As long as at least one get() request is made first, subsequent post() requests would work.  I thought this would be most elegant, as an application doesn't have to be aware of anything happening under the hood - everything would just work.
But I couldn't find a way to update the axios.create() above ...

I had a play with changing the function MerakiDashboard(apiKey) to function MerakiDashboard(apiKey,baseURL).  Then it doesn't break anything that is using it.  It didn't seem very elegant though.
Basically the application using it would make a call to dashboard.organizations.list(), then search for the organisation you want, extract out the url, and then create/replace the existing "dashboard" object but this time include the new URL.  Not so pretty.

**Comment by pdath** [comment]
More specifically, the issue is that axios uses follow-redirects.

It seems that the module follow-redirects deliberatly limits the use of http methods to "read" types.  It doesn't consider POST to be "safe" as it is a "write" method so is changing it on purpose to a GET.

var SAFE_METHODS = { GET: true, HEAD: true, OPTIONS: true, TRACE: true };

[https://github.com/follow-redirects/follow-redirects/blob/master/index.js#L11](https://github.com/follow-redirects/follow-redirects/blob/master/index.js#L11])

**Comment by tejashah88** [comment]
This change could be a breaking change by the looks of it, and the nature of the fix needed would make the library unstable for the time being. There might be a way to find that url from the Meraki Dashboard itself, since that url has some information on the org and other things.

For changing the base url, it might just be a matter of calling `axios.create` again, which seems reasonably easy.

**Comment by pdath** [comment]
Just trying to learn more about this.

I see a lot of examples simply do a:
const axios = require('axios');
And they configure defaults with something like:
axios.defaults.baseURL = 'https://api.example.com';

Is there a reason why you create a seperate instance with axios.create()?  I suspect if the seperate instance was removed then axios.defaults.baseURL could just be updated directly (untested).

**Comment by tejashah88** [comment]
I didn't have a particular reason at the time, except maybe to enforce a sense of immutability and not deal with default or global variables and the bugs that can be created with them. It also would make implementing the base url change fix as a matter of replacing the axios instance with a new one.
